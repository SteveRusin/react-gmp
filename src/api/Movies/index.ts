export { MovieGenre } from './Movie.models';
export type { Movie } from './Movie.models';
export { fetchMovies } from './Movies.mock';
