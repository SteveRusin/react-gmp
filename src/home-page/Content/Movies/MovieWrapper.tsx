import styled from 'styled-components';
import { transparentize } from 'polished';

export const MovieWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 320px;
  gap: 13px;
  cursor: pointer;

  .movie-description-container {
    display: flex;
    justify-content: space-between;
  }

  .movie-description-heading {
    display: flex;
    flex-direction: column;
    gap: 8px;
  }

  .movie-title {
    font-size: 1.8rem;
    opacity: 0.7;
  }

  .movie-genre {
    font-size: 1.4rem;
    opacity: 0.5;
  }

  .movie-release-date {
    align-self: flex-start;
    font-size: 1.4rem;
    opacity: 0.7;
    padding: 0.25em 0.5em;
    border: 1px solid ${({theme}) => transparentize(0.5, theme.palette.textPrimary)};
    border-radius: 4px;
  }
`;
