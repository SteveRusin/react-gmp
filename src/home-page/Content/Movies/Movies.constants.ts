import { MenuItems } from '@shared';

export const enum MenuItemsIds {
  EditMovie,
  DeleteMovie,
}

export const MENU_ITEMS: MenuItems<MenuItemsIds>[] = [
  {
    id: MenuItemsIds.EditMovie,
    name: 'Edit',
  },
  {
    id: MenuItemsIds.DeleteMovie,
    name: 'Delete',
  },
];

