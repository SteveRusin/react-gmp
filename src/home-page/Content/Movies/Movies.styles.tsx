import styled from 'styled-components';
import { transparentize } from 'polished';
import { KebabHorizontal } from '@styled-icons/octicons';

export const Wrapper = styled.div`
  display: grid;
  justify-content: space-between;
  gap: 3rem;
  grid-template-columns: repeat(auto-fill, 320px);
`;

export const MovieWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 320px;
  gap: 13px;
  cursor: pointer;
  position: relative;

  .movie-description-container {
    display: flex;
    justify-content: space-between;
  }

  .movie-description-heading {
    display: flex;
    flex-direction: column;
    gap: 8px;
  }

  .movie-title {
    font-size: 1.8rem;
    opacity: 0.7;
  }

  .movie-genre {
    font-size: 1.4rem;
    opacity: 0.5;
  }

  .movie-release-date {
    align-self: flex-start;
    font-size: 1.4rem;
    opacity: 0.7;
    padding: 0.25em 0.5em;
    border: 1px solid
      ${({ theme }) => transparentize(0.5, theme.palette.textPrimary)};
    border-radius: 4px;
  }

  &:hover,
  &:focus-within {
    .movie-menu-trigger {
      opacity: 1;
      transition: opacity 100ms linear;
    }
  }

  .movie-menu {
    position: absolute;
    top: 0;
    right: 0;
    transform: translate(-50%, 50%);
  }
`;

export const MovieContextMenuTrigger = styled(KebabHorizontal)`
  color: #fff;
  padding: 7px;
  height: 36px;
  width: 36px;
  background-color: ${({ theme }) => theme.palette.accent};
  border-radius: 100%;
  opacity: 0;
  transform: rotate(0.25turn);
`;
