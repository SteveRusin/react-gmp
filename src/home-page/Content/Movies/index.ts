export { MoviesWrapper } from './MoviesWrapper';
export { MenuItemsIds } from './Movies.constants';
