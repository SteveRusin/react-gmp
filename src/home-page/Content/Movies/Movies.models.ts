import { Movie } from '@api/Movies';

export interface MoviesProps {
  movies: Movie[]
}
