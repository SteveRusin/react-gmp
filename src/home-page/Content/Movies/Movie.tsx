import { FC } from 'react';

import { Movie as MovieModel } from '@api/Movies';

import { MovieWrapper } from './MovieWrapper';

export const Movie: FC<MovieModel> = ({
  coverUrl,
  genre,
  releaseDate,
  title,
}) => {
  return (
    <MovieWrapper tabIndex={0}>
      <img className='movie-cover' src={coverUrl} alt='movie cover' />
      <div className='movie-description-container'>
        <div className='movie-description-heading'>
          <span className='movie-title'>{title}</span>
          <span className='movie-genre'>{genre}</span>
        </div>
        <span className='movie-release-date'>{releaseDate}</span>
      </div>
    </MovieWrapper>
  );
};
