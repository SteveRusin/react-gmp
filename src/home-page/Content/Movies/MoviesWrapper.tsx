import { FC, useContext, useState } from 'react';

import {
  MovieDetailsContext,
  AddMovieModal,
  DeleteMovie,
  Modal,
} from '@shared';
import { Movie as MovieModel } from '@api/Movies';

import { Movie } from './Movie';
import { MoviesProps } from './Movies.models';
import { Wrapper } from './Movies.styles';
import { MenuItemsIds } from './Movies.constants';

export const MoviesWrapper: FC<MoviesProps> = ({ movies }) => {
  const { showMovieDetails } = useContext(MovieDetailsContext);
  const [editMovieOpen, setEditMovieOpen] = useState(false);
  const [deleteMovieOpen, setDeleteMovieOpen] = useState(false);
  const [targetMovie, setTargetMovie] = useState<MovieModel | null>(null);
  const showEditModal = editMovieOpen && targetMovie;
  const showDeleteModal = deleteMovieOpen && targetMovie;

  return (
    <>
      <Wrapper>
        {movies.map((movie) => (
          <Movie
            onMenuItemClick={(id) => {
              if (id === MenuItemsIds.EditMovie) {
                setEditMovieOpen(true);
              } else if (id === MenuItemsIds.DeleteMovie) {
                setDeleteMovieOpen(true);
              }
              setTargetMovie(movie);
            }}
            onMovieClick={showMovieDetails}
            movie={movie}
            key={movie.id}
          />
        ))}
      </Wrapper>
      {showDeleteModal && (
        <Modal>
          <DeleteMovie
            onClose={() => {
              setTargetMovie(null);
              setDeleteMovieOpen(false);
            }}
            onConfirm={() => null}
          />
        </Modal>
      )}
      {showEditModal && (
        <Modal>
          <AddMovieModal
            movie={targetMovie}
            onClose={() => {
              setTargetMovie(null);
              setEditMovieOpen(false);
            }}
            onSubmit={() => {
              // do nothing for now
              setTargetMovie(null);
              setEditMovieOpen(false);
            }}
          />
        </Modal>
      )}
    </>
  );
};
