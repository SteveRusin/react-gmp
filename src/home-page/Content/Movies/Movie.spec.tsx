import { Movie as MovieModel, MovieGenre } from '@api/Movies';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Movie } from './Movie';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@theme';
import { MovieProps } from './Movies.models';

function renderComponent(props: MovieProps) {
  return (
    <ThemeProvider theme={defaultTheme}>
      <Movie {...props} />
    </ThemeProvider>
  );
}

describe('Movie Component', () => {
  let onMenuItemClick: jest.Mock<void, unknown[]>;
  let onMovieClick: jest.Mock<void, unknown[]>;
  let movieMock: MovieModel;

  beforeEach(() => {
    onMenuItemClick = jest.fn();
    onMovieClick = jest.fn();
    movieMock = {
      id: 1,
      coverUrl: 'url',
      description: 'description',
      genre: [MovieGenre.ActionAdventure],
      rating: 5,
      releaseDate: '2021',
      runtime: 120,
      title: 'Movie Title',
    };
  });

  it('should render correctly', () => {
    const { asFragment } = render(
      renderComponent({
        onMenuItemClick,
        onMovieClick,
        movie: movieMock,
      })
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should emit clicked move to parent', () => {
    render(
      renderComponent({
        onMenuItemClick,
        onMovieClick,
        movie: movieMock,
      })
    );

    const movieWrapper = screen.getByTestId('movie-wrapper');

    userEvent.click(movieWrapper);

    expect(onMovieClick).toHaveBeenCalledWith(movieMock);
  });
});
