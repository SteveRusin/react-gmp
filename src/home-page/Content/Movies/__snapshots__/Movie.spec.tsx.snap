// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-giImIA ikMeFL"
    data-testid="movie-wrapper"
    tabindex="0"
  >
    <img
      alt="movie cover"
      class="sc-fKFxtB duVGWx"
      src="url"
    />
    <div
      class="movie-description-container"
    >
      <div
        class="movie-description-heading"
      >
        <span
          class="movie-title"
        >
          Movie Title
        </span>
        <span
          class="movie-genre"
        >
          Adventure
        </span>
      </div>
      <span
        class="movie-release-date"
      >
        2021
      </span>
    </div>
    <div
      class="sc-bBXrwG movie-menu"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="StyledIconBase-ea9ulj-0 iuNRof sc-ezredP dFmsqn movie-menu-trigger"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 9a1.5 1.5 0 1 0 0-3 1.5 1.5 0 0 0 0 3zM1.5 9a1.5 1.5 0 1 0 0-3 1.5 1.5 0 0 0 0 3zm13 0a1.5 1.5 0 1 0 0-3 1.5 1.5 0 0 0 0 3z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
